ProtonVPN IP Leakage (#65)
PrivSec-dev/privsec.dev: content/apps/ProtonVPN IP Leakage on Linux and Workaround.md

---
title: "ProtonVPN IP Leakage on Linux and Workaround"
date: 2022-10-08
tags: ['Applications', 'Linux', 'Qubes OS', 'Privacy']
author: Tommy
---

**Before We Start**... 

I sent Proton an email regarding this issue in late August 2022 and was told they are working on fixing it, though it will take some time as it requires some architectural changes in how the killswitch works.

## The Leak

Ideally, when implementing a killswitch, a VPN client should drop all connections on non-VPN interfaces except when the connection is to the VPN provider's servers. This is necessary to prevent accidental leaks, at least by unprivileged applications. Unfortunately, the ProtonVPN client does not currently do this.

> Effectively, any application that binds to the public interface on your Linux system will expose your actual IP address when they connect to the internet regardless of whether you have the killswitch on, off, or "permanently" (persistently) enabled. This is problematic, especially for certain applications like Torrent clients, as they tend to use whatever interfaces they have access to (rather than just the default one) to connect to the internet

Reviewer: ```suggestion
Effectively, any application that binds to the connected physical interface (as opposed to the VPN's virtual interface) on your Linux system will expose your actual IP address, regardless of the killswitch state. This is problematic, especially for certain applications like Torrent clients, as they tend to use whatever interfaces they can access (rather than just the default one) to connect to the internet.
```

> .

Reviewer: ```suggestion

```

You can check this with `curl`: 

```bash

> curl --interface your_public_interface_name https://ipinfo.io

Reviewer: ```suggestion
curl --interface <physical interface> https://ipinfo.io
```

```

This will return your actual IP address.

## The Workaround

### Qubes OS

> On Qubes OS, you generally will not have much of a problem if you use the ProtonVPN client in a ProxyVM. While the same issue still exists within the ProxyVM itself, it is unlikely to happen as you should not be running any other applications in the same qubes anyways. Apps in an AppVM cannot bind to the public interface of the ProxyVM, and I have not observed any leaks from an AppVM behind a ProtonVPN ProxyVM.

Reviewer: ```suggestion
On Qubes OS, you generally should not have an issue if you use the ProtonVPN client in a ProxyVM. While the same issue still exists within the ProxyVM itself, it is unlikely to manifest as you should not be running any other applications in the same Qube anyways, and apps in an AppVM cannot bind to the public interface of the ProxyVM. I have not observed any leaks from an AppVM behind a ProtonVPN ProxyVM.
```

### General Linux Distributions

> On a general Linux distribution, the workaround is to configure OpenVPN manually and setup a killswitch yourself.

Reviewer: ```suggestion
On a general Linux distribution, the workaround is to configure OpenVPN or WireGuard manually and setup a killswitch yourself.
```

Author: Urh, we should probably make a seperate PR for this, as it would be a rabbit hole. 

For one, we need to adjust the rules for Wireguard and test it (I didn't test with Wireguard at all, just OpenVPN). And two, I am not sure if doing PostUp and PreDown would be better than just adding the rules persistently using UFW or Firewalld.
